### ISSUE TITLE
Remove all uses of `panic!`, `unwrap()`, `expect(...)`, and `unreachable!()`

### ISSUE THREAD

**Issue by bhansconnect** [roc-lang/roc#2046]
We want to use our new error macros instead of `panic!`, `unwrap()`, `expect(...)`, and `unreachable!()`. 

The new error macros both live in `roc_reporting` and can be imported with `use roc_reporting:{internal_error, user_error}`.
Essentially, all cases were the `panic!`/`unwrap()`/`expect(...)` would be a sign of a compiler bug, we want to use `internal_error!` instead. `user_error!` should be used in cases where we eventually want to use `roc_reporting` to print a pretty error message. It is essentially documentation that eventually better errors should be return there. `todo!` should be used for cases were the feature is planned to be implemented but is not yet implemented. `unimplemented!` should be used if the feature is not implemented, and there is not a current plan to implement the feature.

Uses of `unreachable!()` should always be replaced with `internal_error!`. If possible, a message around why the statement is unreachable/what invariant was broken, should be added.

This issue will be fixed when we can enable `clippy::panic`, `clippy::unwrap_used`, `clippy::expect_used`, and `clippy::unreachable` on CI by default.
This command can be used to track how many use cases are left:
```bash
cargo clean && cargo clippy --no-deps -- -W clippy::unwrap_used -W clippy::expect_used -W clippy::panic -W clippy::unreachable 2> >(grep -e "warning: \`panic\`" -e "warning: used" -e "warning: usage of" | wc -l )
```

At the date of creating this issue, we have 1283 cases to go.

**Comment by lukewilliamboswell** [comment]
Progress report
```
$ cargo clean && cargo clippy --no-deps -- -W clippy::unwrap_used -W clippy::expect_used -W clippy::panic -W clippy::unreachable 2> >(grep -e "warning: \`panic\`" -e "warning: used" -e "warning: usage of" | wc -l )
1774
```

**Comment by rtfeldman** [comment]
> As an extra notes, in many cases, it may also be fine to replace panic!, unwrap(), and expect(...) with a ? or returning None/Err(...). These are ways of propagating the error up to the calling function. This should be done whenever it is reasonable for the calling function to expect and handle an option or result. If the calling function couldn't cleanly handle and resolve the issues, please stick to other solutions.

As a quick note, this should definitely be the preferred solution (even though it takes longer) whenever possible! A lot of the worst user experiences in Roc right now are due to compiler panics (e.g. because they block later steps in the compiler, such as error reporting, from running as normal), so anywhere we can replace a panic with telling the user what went wrong and then continuing, that's what we should do!

**Comment by lukewilliamboswell** [comment]
> This issue will be fixed when we can enable clippy::panic, clippy::unwrap_used, clippy::expect_used, and clippy::unreachable on CI by default.

This might be a crazy idea... but could we bulk add a `#[ignore(clippy::panic)]` etc on all the locations in the codebase, and then enable this in clip... so that we prevent any further regression in this area? 

Then the process of fixing this is a matter of replacing the `#[ignore(clippy::panic)]` etc which is reasonably easy to search for.

**Comment by Anton-4** [comment]
Could this be simpler: we can execute the count command (`$ cargo clean && cargo clippy --no-deps -- -W clippy::unwrap_used ... )`) in CI and error if it has increased.

If we use  `#[ignore(clippy::panic)]` in the code, people may be tempted to just copy that

**Comment by rtfeldman** [comment]
I like that idea, Anton!

We could also have a script which errors if the number decreases, so you can make sure to decrease the count.

So the error messages could be:
- If the number is higher than what we expect, say "hey this increased panics by N; here's what you should do instead of panicking"
- If the number is lower than what we expect, say "hey it looks like this decreased the number of panics, which is great! Can you update this number in this place to the new number, so we don't regress this progress?"

Also, it's a bit unfortunately, but things like `unwrap` are fine (and useful for conciseness) in tests - but `grep` doesn't know about the distinction between test code and production code. We could make something fancier to detect that distinction in the future perhaps.

**Comment by shua** [comment]
> Also, it's a bit unfortunately, but things like `unwrap` are fine (and useful for conciseness) in tests - but `grep` doesn't know about the distinction between test code and production code. We could make something fancier to detect that distinction in the future perhaps.

no need for anything fancy, the `cargo clippy` command above ignores test code unless you pass `--tests` flag.

```
$ cargo clippy --no-deps -- -W clippy::unwrap_us
ed -W clippy::expect_used -W clippy::panic -W clippy::unreachab
le 2>&1 |grep -A1 '^\(warning: used\|warning: `panic|warning: u
sage of\)' |paste - - - |grep -v ' --> crates/vendor' | wc -l
1209
$ cargo clippy --tests --no-deps -- -W clippy::u
nwrap_used -W clippy::expect_used -W clippy::panic -W clippy::u
nreachable 2>&1 |grep -A1 '^\(warning: used\|warning: `panic|wa
rning: usage of\)' |paste - - - |grep -v ' --> crates/vendor' |
 wc -l
1635
```

**Comment by Anton-4** [comment]
Correct! Clippy was used in https://github.com/roc-lang/roc/pull/7221 :)
